src/nvim/eval.c

```diff
@@ -8018,26 +8019,47 @@ static void f_bufnr(typval_T *argvars, typval_T *rettv)
     rettv->vval.v_number = -1;
 }
 
-/*
- * "bufwinnr(nr)" function
- */
-static void f_bufwinnr(typval_T *argvars, typval_T *rettv)
+static void buf_win_common(typval_T *argvars, typval_T *rettv, bool get_nr)
 {
-  (void)get_tv_number(&argvars[0]);         /* issue errmsg if type error */
-  ++emsg_off;
+  int error = false;
+  (void)get_tv_number_chk(&argvars[0], &error);  // issue errmsg if type error
+  if (error) {  // the argument has an invalid type
+    rettv->vval.v_number = -1;
+    return;
+  }
+
+  emsg_off++;
```

Comment:
@oni-link An error will be shown if there are no matches or if there are multiple matches for the buffer pattern. For example, if we have buffers `foo1.txt` and `foo2.txt`, `bufwinid('foo')` will show an error. The error is shown in the `buflist_findpat` function.

All this was in the original code, but it might be useful to show these errors to the user, the same as when the function receives an argument of the wrong type.